src/Workspaces/Core/Portable/Serialization/SerializerService_Asset.cs

```diff
@@ -61,12 +63,24 @@ private SourceText DeserializeSourceText(ObjectReader reader, CancellationToken
                 var size = reader.ReadInt64();
 
                 var storage = storage2.AttachTemporaryTextStorage(name, offset, size, encoding, cancellationToken);
-
-                return storage.ReadText(cancellationToken);
+                if (storage is ITemporaryTextStorageWithName storageWithName)
+                {
+                    return new SerializableSourceText(storageWithName);
+                }
+                else
+                {
+                    return new SerializableSourceText(storage.ReadText(cancellationToken));
+                }
             }
 
             Contract.ThrowIfFalse(kind == SerializationKinds.Bits);
-            return SourceTextExtensions.ReadFrom(_textService, reader, encoding, cancellationToken);
+            return new SerializableSourceText(SourceTextExtensions.ReadFrom(_textService, reader, encoding, cancellationToken));
+        }
+
+        private SourceText DeserializeSourceText(ObjectReader reader, CancellationToken cancellationToken)
+        {
+            var serializableSourceText = DeserializeSerializableSourceText(reader, cancellationToken);
+            return serializableSourceText.Text ?? serializableSourceText.Storage!.ReadText(cancellationToken);
         }
```

Comment:
Perhaps move this logic to SerializableSourceText.

Reply:
I'm OK with it in the current location for now (same place it's always lived).